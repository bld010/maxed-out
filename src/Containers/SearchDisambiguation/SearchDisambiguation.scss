.SearchDisambiguation {

  & .committee-disambiguation{
    font-size: 14px;
    text-decoration: none;
    color: black;
  }
}