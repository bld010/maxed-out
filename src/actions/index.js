export const setCurrentCandidate = candidate => ({
  type: 'SET_CURRENT_CANDIDATE',
  candidate: candidate
})

export const setCurrentCommitteeId = id => ({
  type: 'SET_CURRENT_COMMITTEE_ID',
  committee_id: id
})

export const setPacContributions = pac_contributions => ({
  type: 'SET_PAC_CONTRIBUTIONS',
  pacContributions: pac_contributions
})

export const setIndividualContributions = individual_contributions => ({
  type: 'SET_INDIVIDUAL_CONTRIBUTIONS', 
  individualContributions: individual_contributions
})