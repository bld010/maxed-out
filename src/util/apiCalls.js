export const searchCandidateByName = async (searchTerm) => {

  let formattedSearchTerm = searchTerm.toLowerCase().split().join('%20')
  let url = `https://api.open.fec.gov/v1/candidates/search/?sort_null_only=false&sort_nulls_last=false&per_page=100&sort=name&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&sort_hide_null=false&name=${formattedSearchTerm}&page=1`


  try {
    let candidateSearchResult = await fetch(url)
    if (candidateSearchResult.ok === false) {
      throw new Error('There was an error searching for the candidate')
    }
    let results = await candidateSearchResult.json()
    return(results.results)
  } catch (err) {
    throw err
  }
}

export const searchCommitteeById = async (committee_id) => {
  let url = `https://api.open.fec.gov/v1/committee/${committee_id}/?sort=name&sort_null_only=false&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&per_page=100&sort_nulls_last=false&sort_hide_null=true&page=1`
  try {
    let committeeSearchResult = await fetch(url);
    if (committeeSearchResult.ok === false) {
      throw new Error('There was an error finding the requested committee')
    }
    let results = await committeeSearchResult.json();
    return(results.results)
  } catch(err) {
    throw err
  }

}

export const searchCandidateById = async (candidate_id) => {

  let url = `https://api.open.fec.gov/v1/candidates/search/?api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&per_page=50&candidate_id=${candidate_id}&sort_nulls_last=false&sort_hide_null=false&page=1&sort=name&sort_null_only=false`

  try {
    let candidateSearchResult = await fetch(url);
    if (candidateSearchResult.ok === false) {
      throw new Error ('There was an error searching for the candidate')
    }
    let results = await candidateSearchResult.json();
    return results.results
  } catch (err) {
    throw err
  }
}

export const fetchPACContributions = async (committee_id) => {

  let url = `https://api.open.fec.gov/v1/schedules/schedule_a/?sort_null_only=false&committee_id=${committee_id}&min_amount=5000&per_page=50&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&is_individual=false&sort_hide_null=false&sort=contribution_receipt_date&two_year_transaction_period=2020`

  try {
    let pacContributions = await fetch(url);
    if (pacContributions.ok === false) {
      throw new Error ('There was an error getting the PAC contributions')
    }
    let results = await pacContributions.json();
    return results.results
  } catch (err) {
    throw err
  }
}

export const fetchIndividualContributions = async (committee_id) => {

  let url = `https://api.open.fec.gov/v1/schedules/schedule_a/?sort_null_only=false&committee_id=${committee_id}&min_amount=2800&per_page=50&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&is_individual=true&sort_hide_null=false&sort=contribution_receipt_date&two_year_transaction_period=2020`

  try {
    let individualContributions = await fetch(url);
    if (individualContributions.ok === false) {
      throw new Error ('There was an error getting the PAC contributions')
    }
    let results = await individualContributions.json();
    return results.results
  } catch (err) {
    throw err
  }
}