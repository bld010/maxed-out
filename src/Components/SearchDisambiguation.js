import React, { Component } from 'react'
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setPacContributions, setCurrentCandidate, setCurrentCommitteeId } from '../actions/index.js';
import { searchCandidateById } from '../util/apiCalls.js';


export class SearchDisambiguation extends Component {


  generateCommitteeList = () => {
  
    return this.props.candidate.principal_committees.map((committee, index) => {
      const { name, state, committee_id, committee_type_full } = committee
      return (
        <Link key={index} to={`/candidate/${this.props.candidate.candidate_id}/committee/${committee_id}`}>
          <article onClick={() => this.props.setCurrentCommitteeId(committee_id)} className="committee-disambiguation">
            <p>{name} ({state} {committee_type_full})</p>
          </article>
        </Link>
    )})
  }

  componentDidMount = async () => {

    try {
      let candidateSearchResults = await searchCandidateById(this.props.candidate_id)
      this.props.setCurrentCandidate(candidateSearchResults[0]);
      
      console.log(this.props)

    } catch (err) {
      console.error(err.message)
    }
    
    
  }
  
  render() {
    console.log(this.props)
    let committeeList; 

    if (this.props.candidate) {
        committeeList = this.generateCommitteeList()
      }

    return(
      <>
      {/* <Route path='/candidate/:candidate_id/committee/:committee_id' render={} */}
      <p>this is sa test</p>
      {this.props.candidate && <p>Candidate: {this.props.candidate.name}</p>}
      {this.props.candidate && <><div>Select a Campaign Committee</div>{committeeList}</> }
      </>
    )
  }
}



export const mapStateToProps = state => ({
  candidate: state.candidate,
  committee_id: state.committee_id
})

export const mapDispatchToProps = dispatch => ({
  setCurrentCommitteeId: committee_id => dispatch(setCurrentCommitteeId(committee_id)),
  setPacContributions: pac_contributions => dispatch(setPacContributions(pac_contributions)),
  setCurrentCandidate: candidate => dispatch(setCurrentCandidate(candidate)),
})

export default connect(mapStateToProps, mapDispatchToProps)(SearchDisambiguation);
