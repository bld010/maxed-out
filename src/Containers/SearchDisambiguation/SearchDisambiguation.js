import React, { Component } from 'react'
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setCurrentCandidate, setCurrentCommitteeId } from '../../actions/index.js';
import { searchCandidateById } from '../../util/apiCalls.js';
import './SearchDisambiguation.scss';
import PropTypes from 'prop-types';


export class SearchDisambiguation extends Component {

  generateCommitteeList = () => {
    return this.props.candidate.principal_committees.map((committee, index) => {
      const { name, state, committee_id, committee_type_full } = committee
      return (
        <Link 
          key={index} 
          to={`/candidate/${this.props.candidate.candidate_id}/committee/${committee_id}`}>
          <button onClick={() => this.props.setCurrentCommitteeId(committee_id)} className="committee-disambiguation">
            {name} ({state} {committee_type_full})
          </button>
        </Link>
    )})
  }

  componentDidMount = async () => {
    try {
      let candidateSearchResults = await searchCandidateById(this.props.candidate_id)
      this.props.setCurrentCandidate(candidateSearchResults[0]);
    } catch (err) {
      console.error(err.message)
    }
  }
  
  render() {

    let committeeList; 

    if (this.props.candidate) {
        committeeList = this.generateCommitteeList()
      }

    if (committeeList !== undefined && committeeList.length === 0) {
      return (
        <h2>No campaign committees found for {this.props.candidate.name}.</h2>
      )
    }

    return(
      <div className="SearchDisambiguation">
        {this.props.candidate && <p>Candidate: {this.props.candidate.name}</p>}
        {this.props.candidate && <><p>Select a Campaign Committee: </p> {committeeList}</> }
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  candidate: state.candidate,
  committee_id: state.committee_id
})

export const mapDispatchToProps = dispatch => ({
  setCurrentCommitteeId: committee_id => dispatch(setCurrentCommitteeId(committee_id)),
  setCurrentCandidate: candidate => dispatch(setCurrentCandidate(candidate)),
})

export default connect(mapStateToProps, mapDispatchToProps)(SearchDisambiguation);

SearchDisambiguation.propTypes = {
  setCurrentCandidate: PropTypes.func.isRequired,
  setCurrentCommitteeId: PropTypes.func.isRequired,
  candidate_id: PropTypes.string,
  committee_id: PropTypes.string,
  candidate: PropTypes.object
}