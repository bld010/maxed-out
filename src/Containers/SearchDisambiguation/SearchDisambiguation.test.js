import React from 'react';
import { shallow } from 'enzyme';
import { setCurrentCommitteeId, setCurrentCandidate } from '../../actions';
import { SearchDisambiguation, mapStateToProps, mapDispatchToProps } from './SearchDisambiguation';
import { searchCandidateById } from '../../util/apiCalls';

jest.mock('../../util/apiCalls')

describe('SearchDisambiguation', () => {
  let mockCommittee = {
    name: 'Pizza 4 All', 
    state: 'CO', 
    committee_id: 'EFG345', 
    committee_type_full: 'Senate'
}
  let mockId = "ABC789"
  let mockCandidate = {
    name: 'Uno', 
    principal_committees: [
      mockCommittee,
      mockCommittee,
      mockCommittee,
      mockCommittee
    ]
  }

  let wrapper;
  let mockSetCurrentCommitteeId;
  let mockSetCurrentCandidate;

  searchCandidateById.mockImplementation(() => {
    return Promise.resolve([mockCandidate])
  })

  beforeEach(() => {

    mockSetCurrentCommitteeId = jest.fn();
    mockSetCurrentCandidate = jest.fn();

    wrapper = shallow(
      <SearchDisambiguation 
        setCurrentCommitteId={mockSetCurrentCommitteeId}
        setCurrentCandidate={mockSetCurrentCandidate}
        candidate={mockCandidate}
        committee_id={mockId}
      />)
  })

  it('should match the snapshot with correct data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('generateCommitteeList', () => {
    it('returns an array of Link elements for each committee', () => {
   

      let result = wrapper.instance().generateCommitteeList();
      expect(result.length).toEqual(4)
    })
  })

  describe('componentDidMount', () => {
    it('fires searchCandidateById with the candidate_id', async () => {
    
      let wrapper = shallow(
        <SearchDisambiguation 
          setCurrentCommitteId={mockSetCurrentCommitteeId}
          setCurrentCandidate={mockSetCurrentCandidate}
          candidate={mockCandidate}
          committee_id={mockId}
          candidate_id={'092384'}
        />)

      expect(searchCandidateById).toHaveBeenCalledWith('092384')
    })

    it('fires setCurrentCandidate with the candidate object', () => {
      let wrapper = shallow(
        <SearchDisambiguation 
          setCurrentCommitteId={mockSetCurrentCommitteeId}
          setCurrentCandidate={mockSetCurrentCandidate}
          candidate={mockCandidate}
          committee_id={mockId}
          candidate_id={'092384'}
        />)

      expect(mockSetCurrentCandidate).toHaveBeenCalledWith(mockCandidate)
    })
  })

  describe('mapStateToProps', () => {
    it('should return a candidate object and a committee_id', () => {
      const mockState = {
        candidate: { name: 'Dorvid' },
        committee_id: 'ABC123'
      }

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(mockState)
    })
  })

  describe('mapDispatchToProps', () => {

    let mockDispatch;
    let mappedDispatchToProps;

    beforeEach(() => {
      mockDispatch = jest.fn();
      mappedDispatchToProps = mapDispatchToProps(mockDispatch)
    })


    it('should call dispatch on setCurrentCandidate action with committee id', () => {
      const mockId = "UML097";
      const dispatchedSetCurrentCandidateAction = setCurrentCandidate(mockId);

      mappedDispatchToProps.setCurrentCandidate(mockId);

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedSetCurrentCandidateAction)

    })


    it('should call dispatch on setCurrentCommitteeId action with committee id', () => {
      const mockId = "ABC123";
      const dispatchedSetCurrentCommitteeIdAction = setCurrentCommitteeId(mockId);

      mappedDispatchToProps.setCurrentCommitteeId(mockId);

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedSetCurrentCommitteeIdAction)

    })
  })
})