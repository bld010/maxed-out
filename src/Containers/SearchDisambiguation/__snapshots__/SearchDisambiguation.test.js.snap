// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchDisambiguation should match the snapshot with correct data passed in 1`] = `
<div
  className="SearchDisambiguation"
>
  <p>
    Candidate: 
    Uno
  </p>
  <p>
    Select a Campaign Committee: 
  </p>
   
  <Link
    key="0"
    to="/candidate/undefined/committee/EFG345"
  >
    <button
      className="committee-disambiguation"
      onClick={[Function]}
    >
      Pizza 4 All
       (
      CO
       
      Senate
      )
    </button>
  </Link>
  <Link
    key="1"
    to="/candidate/undefined/committee/EFG345"
  >
    <button
      className="committee-disambiguation"
      onClick={[Function]}
    >
      Pizza 4 All
       (
      CO
       
      Senate
      )
    </button>
  </Link>
  <Link
    key="2"
    to="/candidate/undefined/committee/EFG345"
  >
    <button
      className="committee-disambiguation"
      onClick={[Function]}
    >
      Pizza 4 All
       (
      CO
       
      Senate
      )
    </button>
  </Link>
  <Link
    key="3"
    to="/candidate/undefined/committee/EFG345"
  >
    <button
      className="committee-disambiguation"
      onClick={[Function]}
    >
      Pizza 4 All
       (
      CO
       
      Senate
      )
    </button>
  </Link>
</div>
`;
